core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -280,6 +290,135 @@ class ReplicaManager(val config: KafkaConfig,
 
   private var logDirFailureHandler: LogDirFailureHandler = null
 
+  // initially true when using a Raft-based metadata quorum, and may flip-flop thereafter;
+  // always false when using ZooKeeper
+  @volatile private var fenced: Boolean = zkClient.isEmpty
+  stateChangeLogger.info(s"Fenced=$fenced")
+
+  def fence(): Unit = {
+    replicaStateChangeLock synchronized {
+      fenced = true
+      stateChangeLogger.info(s"Fenced")
+    }
+  }
+
+  def unfence(): Unit = {
```

Comment:
There seems to be a bit of duplication in here. I think I'm trying to understand why we don't use the `makeFollowers` and `makeLeaders` helpers.